File: solutions/csharp/collatz-conjecture/1/CollatzConjecture.cs
using System;

public static class CollatzConjecture
{
    public static int Steps(int number)
        => number switch
        {
            <= 0 => throw new ArgumentOutOfRangeException(),
            1 => 0,
            _ when (number & 1) == 0 => Steps(number / 2) + 1,
            _ => Steps(number * 3 + 1) + 1,
        };
}

File: solutions/fsharp/collatz-conjecture/1/CollatzConjecture.fs
module CollatzConjecture

let steps (number: int): int option =
    
    let rec steps_internal (number: int): int =
        match number with
        | 1 -> 0
        | _ when number % 2 = 0 -> (steps_internal(number / 2)) + 1
        | _ -> (steps_internal((number * 3) + 1)) + 1

    if number <= 0 then
        None
    else
        steps_internal number |> Some